### PR TITLE
Added ExceptionDataLayoutRenderer and unit tests for it. 

### PR DESCRIPTION
**Pull request by ErickJeffries**
PR for issue: https://github.com/NLog/NLog/issues/4155

Barrowed some from the Exception unit tests, thank you to those who wrote that.

### CODE REVIEW

File: src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    [ThreadAgnostic]
+    [ThreadSafe]
+    public class ExceptionDataLayoutRenderer : LayoutRenderer, IRawValue
+    {
+        private string _dataKey;
+        /// <summary>
+        /// Gets or sets the Data Key 
+        /// </summary>
+        [DefaultParameter]
+        public string DataKey
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs:60]: I prefer if this propery was renamed from `DataKey` to `Item`, as it matches other dictionary-lookup-logic.

File: src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                }
+            }
+        }
+
+        /// <summary>
+        /// Sanitizes the data key against comma separated values and Nulls
+        /// </summary>
+        /// <param name="dataKey">The data key.</param>
+        /// <returns></returns>
+        private static string SanitizeDataKey(string dataKey)
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs:127]: Curious why `SanitizeDataKey` is needed ?

File: src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        /// <summary>
+        /// Gets or sets whether to render innermost Exception from <see cref="Exception.GetBaseException()"/>
+        /// </summary>
+        [DefaultValue(false)]
+        public bool BaseException { get; set; }
+
+        /// <inheritdoc />
+        bool IRawValue.TryGetRawValue(LogEventInfo logEvent, out object value)
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs:84]: Think you should remove the IRawValue-interface-logic

File: src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            return BaseException ? logEvent.Exception?.GetBaseException() : logEvent.Exception;
+        }
+
+        /// <inheritdoc />
+        protected override void Append(StringBuilder builder, LogEventInfo logEvent)
+        {
+            Exception primaryException = GetTopException(logEvent);
+            if (primaryException != null)
+            {
+                AppendDataItem(builder, primaryException, DataKey);
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs:101]: Think you should remove `protected virtual AppendDataItem` and move the implementation here.

Consider adding a `string Format { get; set; }`-property and use it like this:

            var value = ex.Data[dataKey];
            if (value != null)
            {
                var formatProvider = GetFormatProvider(logEvent);
                builder.AppendFormattedValue(value, Format, formatProvider, ValueFormatter);
            }

**Review by snakefoot — CHANGES_REQUESTED** [review]
Very good start. Just need to remove some noise that probably was introduced by copying from the original `ExceptionLayoutRenderer` . Maybe take a look at one of the layoutrenderer used for dictionary-lookups like `GdcLayoutRenderer`.

File: tests/NLog.UnitTests/LayoutRenderers/ExceptionDataLayoutRendererTests.cs [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+                return null;
+            }
+            catch (Exception ex)
+            {
+                ex.Data["1Really.Bad-Boy!"] = "Hello World";
+                return ex;
+            }
+        }
+
+        private int NestedFunc(int recursion, Func<int> innerAction)
```
Comment (snakefoot) [inline tests/NLog.UnitTests/LayoutRenderers/ExceptionDataLayoutRendererTests.cs:272]: This doesn't seem to used for anything

File: tests/NLog.UnitTests/LayoutRenderers/ExceptionDataLayoutRendererTests.cs [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        {
+            public override string Message => throw new Exception("Exception from Message property");
+        }
+
+        /// <summary>
+        /// Get an exception with stacktrace by generating a exception
+        /// </summary>
+        /// <param name="exceptionMessage"></param>
+        /// <returns></returns>
+        private Exception GetExceptionWithStackTrace(string exceptionMessage)
```
Comment (snakefoot) [inline tests/NLog.UnitTests/LayoutRenderers/ExceptionDataLayoutRendererTests.cs:229]: This funky method used to generate exceptions with exotic stacktraces doesn't seem relevant for doing Data-collection-lookups. I can understand the need for nested exceptions to test BaseException-property, but the exception doesn't need to be thrown from a special context to test this. Better with simple arrange + act + assert.

File: src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs (hunk 3 above)
Comment (ErickJeffries) [reply src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs]: done

**Comment by ErickJeffries** [comment]
@snakefoot, your comments were super helpful, I implemented all of them and that greatly simplified things. I did originally start of thinking of inheriting from "ExceptionLayoutRenderer" and then deviated but never really made the jump to just being it's own. I've done that now along with the unit tests being their own rather than a poor copy/paste/edit. This commit feels much better. let me know what you see.

**Review by snakefoot — APPROVED** [review]
Looks much better, thank you for the quick update.

**Comment by snakefoot** [comment]
@ErickJeffries Are you okay with the changes being included with NLog 5 when released, or are you in a hurry?

**Comment by ErickJeffries** [comment]
@snakefoot I'm not in a hurry, NLog 5 is fine with me. I have used NLog for years and I wanted to start participating in Open Source Development so I picked this project as the one I'd start with :)

File: src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.Text;
+    using NLog.Common;
+    using NLog.Config;
+    using NLog.Internal;
+
+    /// <summary>
+    /// Exception information provided through
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs:45]:
```suggestion
    /// Render information of <see cref="Exception.Data" />
```

File: src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.Text;
+    using NLog.Common;
+    using NLog.Config;
+    using NLog.Internal;
+
+    /// <summary>
+    /// Exception information provided through 
+    /// a call to one of the Logger.*Exception() methods.
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs:46]:
```suggestion
    /// for the exception passed to the logger call
```

**Review by 304NotModified — CHANGES_REQUESTED** [review]
The *Exception methods are obsolete for a while, I see we need to update also 

https://github.com/NLog/NLog/blob/dev/src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs

File: src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs (hunk 8 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/ExceptionDataLayoutRenderer.cs:46]: better wording allowed :)

**Comment by snakefoot** [comment]
Closing and reopening to rebuild Travis

**Comment by snakefoot** [comment]
@ErickJeffries Thank you for the contribution.

**Comment by snakefoot** [comment]
Added wiki: https://github.com/NLog/NLog/wiki/ExceptionData-Layout-Renderer

**Pull request merged**